pyuvdata/uvdata/uvdata.py

```diff
@@ -5563,14 +5570,178 @@ def compress_by_redundancy(self, tol=1.0, inplace=True, keep_all_metadata=True):
+                            "missing those times."
+                        )
+                        continue
+
+                    regular_orientation = np.array(
+                        [time_ind for time_ind in gp if time_ind < len(group_times)],
+                        dtype=np.int,
+                    )
+                    regular_inds = group_inds[np.array(regular_orientation)]
+                    conj_orientation = np.array(
+                        [time_ind for time_ind in gp if time_ind >= len(group_times)],
+                        dtype=np.int,
+                    )
+                    conj_inds = np.array(conj_group_inds[np.array(conj_orientation)])
+                    # check that the integration times are all the same
+                    int_times = np.concatenate(
+                        (
+                            self.integration_time[regular_inds],
+                            self.integration_time[conj_inds],
+                        )
+                    )
+                    if not np.all(
+                        np.abs(int_times - new_obj.integration_time[obj_time_ind])
+                        < new_obj._integration_time.tols[1]
+                    ):
+                        warnings.warn(
+                            "Integrations times are not identical in a redundant "
+                            "group. Averaging anyway but this may cause unexpected "
+                            "behavior."
+                        )
+
+                    if not self.metadata_only:
+                        vis_to_avg = np.concatenate(
+                            (
+                                self.data_array[regular_inds, :, :, :],
+                                np.conj(self.data_array[conj_inds, :, :, :]),
+                            )
+                        )
+                        nsample_to_avg = np.concatenate(
+                            (
+                                self.nsample_array[regular_inds, :, :, :],
+                                self.nsample_array[conj_inds, :, :, :],
+                            )
+                        )
+                        flags_to_avg = np.concatenate(
+                            (
+                                self.flag_array[regular_inds, :, :, :],
+                                self.flag_array[conj_inds, :, :, :],
+                            )
+                        )
+                        # if all data is flagged, average it all as if it were not
+                        if np.all(flags_to_avg):
+                            mask = np.zeros_like(flags_to_avg)
+                        else:
+                            mask = flags_to_avg
+
+                        vis_to_avg = np.ma.masked_array(vis_to_avg, mask=mask)
+
+                        nsample_to_avg = np.ma.masked_array(nsample_to_avg, mask=mask)
+
+                        avg_vis = np.average(vis_to_avg, weights=nsample_to_avg, axis=0)
```

Comment:
Just want to verify  we are getting the behavior here that we want. When this call is invoked without explicitly calling the numpy.ma.average function it will count the weights of the masked items towards the total weights. If you only want to include the weights of the unmasked items we would want to us the `ma` module in numpy.

Ah there is also a `returned` keyword in the average function that can return the summed weights if desired.

See this example if this was confusing. I added a suggestion in case you wanted to move to the `ma` module's average. 
```
(py38) matthew@bahamut [9:00:27] [~] 
-> % ipython       
Python 3.8.1 (default, Jan  8 2020, 22:29:32) 
Type 'copyright', 'credits' or 'license' for more information
IPython 7.11.1 -- An enhanced Interactive Python. Type '?' for help.

In [1]: import numpy as np                                                                               

In [2]: this = np.arange(10)                                                                             

In [4]: this = np.ma.masked_array(this, mask=[1 if ind % 2 ==0 else 0 for ind in range(10)])             

In [6]: this                                                                                             
Out[6]: 
masked_array(data=[--, 1, --, 3, --, 5, --, 7, --, 9],
             mask=[ True, False,  True, False,  True, False,  True, False,
                    True, False],
       fill_value=999999)

In [7]: np.average(this)                                                                                 
Out[7]: 5.0

In [8]: this.mean()                                                                                      
Out[8]: 5.0

In [9]: np.average(this, weights=np.arange(10))                                                          
Out[9]: 3.6666666666666665

In [10]: np.sum(this*np.arange(10))/np.sum(np.arange(10))                                                
Out[10]: 3.6666666666666665

In [11]: np.ma.average(this, weights=np.arange(10))                                                      
Out[11]: 6.6

```
```suggestion
                        avg_vis = np.ma.average(vis_to_avg, weights=nsample_to_avg, axis=0)
```

Reply:
oh, good catch. I think we do want the `ma` one here. I also want to go check the time & frequency averaging to make sure they're doing it correctly. I also want to add a check to catch this, let me think about how to do that.

Reply:
Ha! so I made a test to catch this and it passed. Then I messed around on the command line and discovered that we're actually doing the right thing (accidentally) because we are passing a masked array as the weights, which gets us the correct behavior even when we use `np.average`. I think this should be changed or at least commented to make sure we don't confuse anyone in the future though.

My demonstration:
```
In [1]: import numpy as np                                                                                                    

In [2]: data = [1, 0, 0, 0, 0]                                                                                                

In [3]: flags = [False, True, True, True, True]                                                                               

In [4]: nsample = [1, 1, 1, 1, 1]                                                                                             

In [5]: masked_data = np.ma.masked_array(data = data, mask = flags)                                                           

In [6]: masked_data                                                                                                           
Out[6]: 
masked_array(data=[1, --, --, --, --],
             mask=[False,  True,  True,  True,  True],
       fill_value=999999)

In [7]: np.average(masked_data)                                                                                               
Out[7]: 1.0

In [8]: np.average(masked_data, weights=nsample)                                                                              
Out[8]: 0.2

In [9]: np.ma.average(masked_data, weights=nsample)                                                                           
Out[9]: 1.0

In [10]: masked_nsample = np.ma.masked_array(data=nsample, mask=flags)                                                        

In [11]: np.average(masked_data, weights=masked_nsample)                                                                      
Out[11]: 1.0

In [12]: np.ma.average(masked_data, weights=masked_nsample)                                                                   
Out[12]: 1.0
```

Reply:
ah interesting. nice catch there. Think of the terror if the weights and data arrays had different masks going in :scream:

Reply:
The horror! I don't even know what should happen in that case.

Reply:
I'll commit my test that checks this and change to using `np.ma.average` to make sure there's no confusion in the future.

Reply:
I just checked the time & frequency averaging, they look ok.